File: typescript/examples/trigger_orders.ts
#!/usr/bin/env npx ts-node
// @ts-nocheck
/**
 * Trigger Orders Example
 *
 * Stop loss and take profit orders.
 *
 * Requires: PRIVATE_KEY environment variable
 */

import { HyperliquidSDK, Side } from '@quicknode/hyperliquid-sdk';

const PRIVATE_KEY = process.env.PRIVATE_KEY;

if (!PRIVATE_KEY) {
  console.error("Set PRIVATE_KEY environment variable");
  console.error("Example: export PRIVATE_KEY='0x...'");
  process.exit(1);
}

async function main() {
  const sdk = new HyperliquidSDK(undefined, { privateKey: PRIVATE_KEY });
  const mid = await sdk.getMid("BTC");
  console.log(`BTC mid: $${mid.toLocaleString()}`);

  // Stop loss order (market) - triggers when price falls below stop price
  // No limitPrice means market order when triggered
  // const result = await sdk.stopLoss("BTC", {
  //   size: 0.001,
  //   triggerPrice: mid * 0.95,  // 5% below current
  // });
  // console.log(`Stop loss (market): ${result}`);

  // Stop loss order (limit) - triggers and places limit order at limitPrice
  // const result = await sdk.stopLoss("BTC", {
  //   size: 0.001,
  //   triggerPrice: mid * 0.95,
  //   limitPrice: mid * 0.94,
  // });
  // console.log(`Stop loss (limit): ${result}`);

  // Take profit order (market) - triggers when price rises above trigger
  // const result = await sdk.takeProfit("BTC", {
  //   size: 0.001,
  //   triggerPrice: mid * 1.05,  // 5% above current
  // });
  // console.log(`Take profit (market): ${result}`);

  // Take profit order (limit)
  // const result = await sdk.takeProfit("BTC", {
  //   size: 0.001,
  //   triggerPrice: mid * 1.05,
  //   limitPrice: mid * 1.06,
  // });
  // console.log(`Take profit (limit): ${result}`);

  // For buy-side stop/TP (e.g., closing a short position), use side=Side.BUY
  // const result = await sdk.stopLoss("BTC", { size: 0.001, triggerPrice: mid * 1.05, side: Side.BUY });

  console.log("\nTrigger order methods available:");
  console.log("  sdk.stopLoss(asset, { size, triggerPrice, limitPrice?, side? })");
  console.log("  sdk.takeProfit(asset, { size, triggerPrice, limitPrice?, side? })");
  console.log("  sdk.triggerOrder(TriggerOrder(...))");
  console.log("\nNote: Omit limitPrice for market orders when triggered");
}

main().catch(console.error);

File: python/examples/trigger_orders.py
#!/usr/bin/env python3
"""
Trigger Orders Example

Stop loss and take profit orders.

Requirements:
    pip install hyperliquid-sdk

Usage:
    export PRIVATE_KEY="0x..."
    python trigger_orders.py
"""

import os
import sys

from hyperliquid_sdk import HyperliquidSDK

PRIVATE_KEY = os.environ.get("PRIVATE_KEY")

if not PRIVATE_KEY:
    print("Error: Set PRIVATE_KEY environment variable")
    print("  export PRIVATE_KEY='0xYourPrivateKey'")
    sys.exit(1)


def main():
    print("Hyperliquid Trigger Orders Example")
    print("=" * 50)

    sdk = HyperliquidSDK(private_key=PRIVATE_KEY)
    mid = sdk.get_mid("BTC")
    print(f"BTC mid: ${mid:,.2f}")

    # Stop loss order (market) - triggers when price falls below stop price
    # No limit_price means market order when triggered
    # result = sdk.stop_loss("BTC", size=0.001, trigger_price=mid * 0.95)
    # print(f"Stop loss (market): {result}")

    # Stop loss order (limit) - triggers and places limit order at limit_price
    # result = sdk.stop_loss("BTC", size=0.001, trigger_price=mid * 0.95, limit_price=mid * 0.94)
    # print(f"Stop loss (limit): {result}")

    # Take profit order (market) - triggers when price rises above trigger
    # result = sdk.take_profit("BTC", size=0.001, trigger_price=mid * 1.05)
    # print(f"Take profit (market): {result}")

    # Take profit order (limit)
    # result = sdk.take_profit("BTC", size=0.001, trigger_price=mid * 1.05, limit_price=mid * 1.06)
    # print(f"Take profit (limit): {result}")

    # For buy-side stop/TP (e.g., closing a short position), use side="buy"
    # result = sdk.stop_loss("BTC", size=0.001, trigger_price=mid * 1.05, side="buy")

    print()
    print("Trigger order methods available:")
    print("  sdk.stop_loss(asset, size, trigger_price, limit_price=None, side=None)")
    print("  sdk.take_profit(asset, size, trigger_price, limit_price=None, side=None)")
    print("  sdk.trigger_order(TriggerOrder(...))")
    print()
    print("Note: Omit limit_price for market orders when triggered")


if __name__ == "__main__":
    main()
